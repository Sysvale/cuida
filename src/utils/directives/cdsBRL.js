export const vCdsBrl = {
	mounted(el, binding) {
		const formatValue = (value) => {
			if (value == 0) return 'R$ 0,00';
			if (!value) return '';

			let onlyNumbers = String(value).replace(/\D/g, '');
			if (!onlyNumbers) return 'R$ 0,00';

			let number = parseFloat(onlyNumbers) / 100;
			return number.toLocaleString('pt-BR', {
				style: 'currency',
				currency: 'BRL',
			});
		};

		let isFormatting = false;

		const inputHandler = (event) => {
			if (isFormatting) return;

			isFormatting = true;

			let currentValue = String(event.target.value || '');

			let onlyNumbers = currentValue.replace(/\D/g, '');
			let number = parseFloat(onlyNumbers) / 100;

			const formattedValue = formatValue(currentValue);
			event.target.value = formattedValue;

			const inputEvent = new Event('input', { bubbles: true });

			Object.defineProperty(inputEvent, 'target', {
				value: { value: number },
				writable: false,
			});

			el.dispatchEvent(inputEvent);
			isFormatting = false;
		};

		el.addEventListener('input', inputHandler);

		el.value = formatValue(binding.value || '');

		el._inputHandler = inputHandler;
	},

	updated(el, binding) {
		const formatValue = (value) => {
			if (!value) return '';

			let onlyNumbers = String(value).replace(/\D/g, '');
			let number = parseFloat(onlyNumbers) / 100;

			return number.toLocaleString('pt-BR', {
				style: 'currency',
				currency: 'BRL',
			});
		};

		el.value = formatValue(binding.value || '');
	},

	unmounted(el) {
		if (el._inputHandler) {
			el.removeEventListener('input', el._inputHandler);
		}
	},
};

export const unmaskBRL = (value) => {
	let sanitizedValue = value.replace('R$', '');
	sanitizedValue = sanitizedValue.replaceAll('.', '');
	sanitizedValue = sanitizedValue.replace(/(.*),(\d{2})/g, '$1.$2');

	return +sanitizedValue;
};
