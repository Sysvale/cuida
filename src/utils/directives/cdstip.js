import tippy from 'tippy.js';
import 'tippy.js/dist/tippy.css';
import 'tippy.js/animations/shift-away-subtle.css';

export default {
	mounted(el, binding) {
		if (!binding.value || (typeof binding.value !== 'string' && !binding.value.content)) {
			return;
		}

		const content = typeof binding.value === 'string' ? binding.value : binding.value.content || '';

		el._tippyInstance = tippy(el, {
			content,
			arrow: false,
			animation: 'shift-away-subtle',
			delay: 100,
			allowHTML: true,
			...(binding.value.options || {}),
		});
	},

	updated(el, binding) {
		if (!binding.value || (typeof binding.value !== 'string' && !binding.value.content)) {
			if (el._tippyInstance) {
				el._tippyInstance.destroy();
				el._tippyInstance = null;
			}
			return;
		}

		const content = typeof binding.value === 'string' ? binding.value : binding.value.content || '';

		if (el._tippyInstance) {
			el._tippyInstance.setProps({
				content,
				...(binding.value.options || {}),
			});
		} else {
			el._tippyInstance = tippy(el, {
				content,
				arrow: false,
				animation: 'shift-away-subtle',
				delay: 100,
				allowHTML: true,
				...(binding.value.options || {}),
			});
		}
	},

	beforeUnmount(el) {
		if (el._tippyInstance) {
			el._tippyInstance.destroy();
			el._tippyInstance = null;
		}
	},
};
