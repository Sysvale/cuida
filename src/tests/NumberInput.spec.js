import { describe, test, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import CdsNumberInput from '../components/NumberInput.vue';
import CdsBaseInput from '../components/BaseInput.vue';
import { vCdsBrl } from '../utils/directives/cdsBRL';


describe('CdsNumberInput', () => {
	test('renders correctly', () => {
		const wrapper = mount(CdsNumberInput, {
			props: {
				id: 'number-input',
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('renders correctly with default props', () => {
		const wrapper = mount(CdsNumberInput, {
			props: {
				id: 'number-input',
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).exists()).toBe(true);
		expect(wrapper.findComponent(CdsBaseInput).props('label')).toBe('Valor');
		expect(wrapper.findComponent(CdsBaseInput).props('placeholder')).toBe('Digite aqui a informação');
	});

	test('renders with a custom label', () => {
		const wrapper = mount(CdsNumberInput, {
			props: {
				id: 'number-input',
				label: 'Custom Label',
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('label')).toBe('Custom Label');
	});

	test('renders with a custom placeholder', () => {
		const wrapper = mount(CdsNumberInput, {
			props: {
				id: 'number-input',
				placeholder: 'Custom Placeholder',
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('placeholder')).toBe('Custom Placeholder');
	});

	test('renders with a required indicator', () => {
		const wrapper = mount(CdsNumberInput, {
			props: {
				id: 'number-input',
				required: true,
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('required')).toBe(true);
	});

	test('renders with a disabled attribute when disabled prop is true', () => {
		const wrapper = mount(CdsNumberInput, {
			props: {
				id: 'number-input',
				disabled: true,
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('disabled')).toBe(true);
	});

	test('renders with a floating label when mobile prop is true', () => {
		const wrapper = mount(CdsNumberInput, {
			props: {
				id: 'number-input',
				mobile: true,
			},
		});

		expect(wrapper.findComponent(CdsBaseInput).props('floatingLabel')).toBe(true);
	});

	test('applies money mask and updates internal state correctly', async () => {
		const wrapper = mount(CdsNumberInput, {
			global: {
				directives: {
					CdsBrl: vCdsBrl, // Registre a diretiva
				},
			},
			props: {
				id: 'number-input',
				money: true,
			},
		});

		await wrapper.vm.$nextTick();

		const input = wrapper.find('input');

		await input.setValue('123456');

		expect(wrapper.vm.model).toMatch(/R\$[\s\u00A0]1\.234,56/);
		expect(wrapper.vm.unmaskedValue).toBe(1234.56);
	});
});