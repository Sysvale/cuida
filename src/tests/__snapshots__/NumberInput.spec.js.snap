// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CdsNumberInput > renders correctly 1`] = `
"<div id="number-input">
  <div data-v-82c69faf="" money="false" linkurl="https://cuida.framer.wiki/" mobile="false"><label data-v-82c69faf="" class="base-input__label" for="cds-base-input-number-number-input"><span data-v-82c69faf="" class="label__content">Valor <!--v-if--><!--v-if--></span>
      <!--v-if-->
    </label>
    <div data-v-82c69faf="" class="base-input base-input--default">
      <!--v-if--><input data-v-82c69faf="" id="number-input" modelvalue="" label="Valor" state="default" type="number" placeholder="Digite aqui a informação" errormessage="Valor inválido" supportingtext="" fluid="false" tooltipicon="info-outline" supportlinkurl="https://cuida.framer.wiki/" floatinglabel="false" lazy="false" name="" class="base-input__field">
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>"
`;
